File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/Conversation/View/MSUnreadView.m
//
//  MSUnreadView.m
//  BlackFireIM
//
//  Created by benny wang on 2021/5/28.
//

#import "MSUnreadView.h"

@implementation MSUnreadView

- (instancetype)initWithFrame:(CGRect)frame
{
    if (self = [super initWithFrame:frame]) {
        [self setupViews];
        [self defaultLayout];
    }
    return self;
}

- (void)setNum:(NSInteger)num
{
    NSString *unReadStr = [[NSNumber numberWithInteger:num] stringValue];
//    if (num > 99){
//        unReadStr = @"99+";
//    }
    _unReadLabel.text = unReadStr;
    self.hidden = (num <= 0? YES: NO);
    [self defaultLayout];
}

- (void)setupViews
{
    _unReadLabel = [[UILabel alloc] init];
    _unReadLabel.text = @"11";
    _unReadLabel.font = [UIFont systemFontOfSize:12];
    _unReadLabel.textColor = [UIColor whiteColor];
    _unReadLabel.textAlignment = NSTextAlignmentCenter;
    [_unReadLabel sizeToFit];
    [self addSubview:_unReadLabel];

    self.layer.cornerRadius = (_unReadLabel.frame.size.height + 2 * 2)/2.0;
    [self.layer masksToBounds];
    self.backgroundColor = [UIColor redColor];
    self.hidden = YES;
}

- (void)defaultLayout
{
    [_unReadLabel sizeToFit];
    CGFloat width = _unReadLabel.frame.size.width + 2 * 4;
    CGFloat height =  _unReadLabel.frame.size.height + 2 * 2;
    if(width < height){
        width = height;
    }
    self.bounds = CGRectMake(0, 0, width, height);
    _unReadLabel.frame = self.bounds;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    if (@available(iOS 11.0, *)){
        //Here is a workaround on iOS 11 UINavigationBarItem init with custom view, position issue
        UIView *view = self;
        while (![view isKindOfClass:[UINavigationBar class]] && [view superview] != nil)
        {
            view = [view superview];
            if ([view isKindOfClass:[UIStackView class]] && [view superview] != nil)
            {
                    CGFloat margin = 40.0f;
                        //margin = 4.0f;
                    [view.superview addConstraint:[NSLayoutConstraint constraintWithItem:view
                                                                                    attribute:NSLayoutAttributeLeading
                                                                                    relatedBy:NSLayoutRelationEqual
                                                                                    toItem:view.superview
                                                                                    attribute:NSLayoutAttributeLeading
                                                                                    multiplier:1.0
                                                                                    constant:margin]];
                break;
            }
        }
    }
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/Conversation/View/MSUnreadView.swift
//
//  MSUnreadView.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/8/10.
//

import UIKit

open class MSUnreadView: UIView {

    public var unReadLabel: UILabel!
    
    public func setNum(num: Int) {
        
        let unReadStr = String(num)
        unReadLabel.text = unReadStr
        isHidden = num <= 0 ? true : false
        defaultLayout()
    }

    public override init(frame: CGRect) {
        super.init(frame: frame)
        
        unReadLabel = UILabel()
        unReadLabel.text = "99"
        unReadLabel.font = .systemFont(ofSize: 12)
        unReadLabel.textColor = .white
        unReadLabel.textAlignment = .center
        unReadLabel.sizeToFit()
        addSubview(unReadLabel)
        
        layer.cornerRadius = (unReadLabel.height + 2 * 2) * 0.5
        layer.masksToBounds = true
        backgroundColor = .red
        isHidden = true
        
        defaultLayout()
    }
    
    private func defaultLayout() {
        
        unReadLabel.sizeToFit()
        var width = unReadLabel.width + 2 * 4
        let height = unReadLabel.height + 2 * 2
        if width < height {
            width = height
        }
        bounds = CGRect(x: 0, y: 0, width: width, height: height)
        unReadLabel.frame = bounds
    }
    
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    open override func layoutSubviews() {
        super.layoutSubviews()
        var view: UIView = self
        while view.isKind(of: UINavigationBar.self) == false && view.superview != nil {
            view = view.superview!
            if view.isKind(of: UIStackView.self) && view.superview != nil {
                let margin: CGFloat = 40
                view.superview?.addConstraint(NSLayoutConstraint(item: view, attribute: .leading, relatedBy: .equal, toItem: view.superview, attribute: .leading, multiplier: 1.0, constant: margin))
            }
        }
    }
}
